modules/afmongodb/afmongodb-legacy-uri.c

```diff
@@ -0,0 +1,301 @@
+        {
+          msg_error("Cannot parse address",
+                    evt_tag_str("primary", g_list_nth_data(self->servers, 0)),
+                    evt_tag_str("driver", self->super.super.super.id));
+          return FALSE;
+        }
+      _append_host(&self->recovery_cache, self->address, 0);
+    }
+  return TRUE;
+}
+
+static gboolean
+_append_servers(GString *uri_str, const GList *recovery_cache, gboolean *have_uri)
+{
+  const GList *iterator = recovery_cache;
+  *have_uri = FALSE;
+  gboolean have_path = FALSE;
+  do
+    {
+      const MongoDBHostPort *hp = (const MongoDBHostPort *)iterator->data;
+      if (hp->port)
+        {
+          *have_uri = TRUE;
+          if (have_path)
+            {
+              msg_warning("Cannot specify both a domain socket and address");
+              return FALSE;
+            }
+          g_string_append_printf(uri_str, "%s:%d", hp->host, hp->port);
+        }
+      else
+        {
+          have_path = TRUE;
+          if (*have_uri)
+            {
+              msg_warning("Cannot specify both a domain socket and address");
+              return FALSE;
+            }
+          g_string_append_printf(uri_str, "%s", hp->host);
+        }
+      iterator = iterator->next;
+      if (iterator)
+        g_string_append_printf(uri_str, ",");
+    }
+  while (iterator);
+  return TRUE;
+}
+
+static gboolean
+_check_auth_options(MongoDBDestDriver *self)
+{
+  if (self->user || self->password)
+    {
+      if (!self->user || !self->password)
+        {
+          msg_error("Neither the username, nor the password can be empty");
+          return FALSE;
+        }
+    }
+  return TRUE;
+}
+
+gboolean
+_build_uri_from_legacy_options(MongoDBDestDriver *self)
+{
+    if (!_append_legacy_servers(self))
+      return FALSE;
+
+    self->uri_str = g_string_new("mongodb://");
+    if (!self->uri_str)
+      return FALSE;
+
+    if (self->user && self->password)
+      {
+        g_string_append_printf(self->uri_str, "%s:%s@", self->user, self->password);
+      }
+
+    if (!self->recovery_cache)
+      {
+        msg_error("Error in host server list", evt_tag_str("driver", self->super.super.super.id));
+        return FALSE;
+      }
+
+    gboolean have_uri;
+    if (!_append_servers(self->uri_str, self->recovery_cache, &have_uri))
+      return FALSE;
+
+    if (have_uri)
+      g_string_append_printf(self->uri_str, "/%s", self->db);
+
+    if (self->safe_mode)
+      g_string_append_printf(self->uri_str,
+                             "?wtimeoutMS=%d",
+                             SOCKET_TIMEOUT_FOR_MONGO_CONNECTION_IN_MILLISECS);
+    else
+      g_string_append(self->uri_str, "?w=0&safe=false");
+
+    g_string_append_printf(self->uri_str,
+                           "&socketTimeoutMS=%d&connectTimeoutMS=%d",
+                           SOCKET_TIMEOUT_FOR_MONGO_CONNECTION_IN_MILLISECS,
+                           SOCKET_TIMEOUT_FOR_MONGO_CONNECTION_IN_MILLISECS);
+
```

Comment:
Looking at these options, how they get encoded into an URL, I am not sure it is a good idea to keep the uri the preferred interface for mongo options.

How is it accomplished in other applications?